proxy/src/proxy.rs

```diff
@@ -22,12 +54,22 @@ pub fn thread_main(
         println!("accepted connection from {}", peer_addr);
         socket.set_nodelay(true).unwrap();
 
+        // TODO Use a threadpool instead. Maybe use tokio's threadpool by
+        //      spawning a future into its runtime. Tokio's JoinError should
+        //      allow us to handle cleanup properly even if the future panics.
         thread::Builder::new()
             .name("Proxy thread".into())
             .spawn(move || {
                 if let Err(err) = proxy_conn_main(state, socket) {
                     println!("error: {}", err);
                 }
+
+                // Clean up CANCEL_MAP.
+                THREAD_CANCEL_KEY_DATA.with(|cell| {
+                    if let Some(cancel_key_data) = cell.get() {
+                        CANCEL_MAP.lock().remove(&cancel_key_data);
```

Comment:
Hm what if client closes the connection and then cancels? We'll close our connection to the backend, but not sure if the backend would notice and stop query execution. We could just send a cancel request here ourselves? I'll think about it and address it...

Reply:
> not sure if the backend would notice and stop query execution

Actually it wouldn't notice with or without proxy. There were few suggestions on how to fix that (periodically check socket with MSG_PEEK or FIONREAD) but nothing get actually merged.

Reply:
(but it would notice after the query is done -- when backend will try to read next query from the socket)

Reply:
> Actually it wouldn't notice with or without proxy. There were few suggestions on how to fix that (periodically check socket with MSG_PEEK or FIONREAD) but nothing get actually merged.

In PG 14 `client_connection_check_interval` was added, see https://www.postgresql.org/docs/current/runtime-config-connection.html#RUNTIME-CONFIG-CONNECTION-SETTINGS

IIUC, this is there to address exactly the aforementioned issue. So without proxy Postgres should handle it properly (and if proxy will close the connection to PG, this should work too). We don't configure `client_connection_check_interval` now, though